backtracking/sudoku.py

```diff
@@ -51,24 +54,21 @@ def is_safe(grid: Matrix, row: int, column: int, n: int) -> bool:
         if grid[row][i] == n or grid[i][column] == n:
             return False
 
-    for i in range(3):
-        for j in range(3):
-            if grid[(row - row % 3) + i][(column - column % 3) + j] == n:
-                return False
-
-    return True
+    return all(
+        grid[(row - row % 3) + i][(column - column % 3) + j] != n
+        for i, j in itertools.product(range(3), range(3))
+    )
 
 
 def find_empty_location(grid: Matrix) -> tuple[int, int] | None:
     """
     This function finds an empty location so that we can assign a number
     for that particular row and column.
     """
-    for i in range(9):
-        for j in range(9):
-            if grid[i][j] == 0:
-                return i, j
-    return None
+    return next(
+        ((i, j) for i, j in itertools.product(range(9), range(9)) if grid[i][j] == 0),
```

Comment:
```suggestion
        ((i, j) for i in range(9) for j in range(9) if grid[i][j] == 0),
```
This should do the same thing without needed an imported function